oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/NimbusJwtClientAuthenticationCustomizer.java

```diff
@@ -0,0 +1,216 @@
+ * Using Assertions for Client Authentication</a>
+ * @see <a target="_blank" href="https://connect2id.com/products/nimbus-jose-jwt">Nimbus
+ * JOSE + JWT SDK</a>
+ */
+public final class NimbusJwtClientAuthenticationCustomizer<T extends AbstractOAuth2AuthorizationGrantRequest>
+		implements OAuth2AuthorizationGrantRequestEntityConverter.Customizer<T> {
+
+	private static final String INVALID_KEY_ERROR_CODE = "invalid_key";
+
+	private static final String INVALID_ALGORITHM_ERROR_CODE = "invalid_algorithm";
+
+	private static final String CLIENT_ASSERTION_TYPE_VALUE = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
+
+	private final Function<ClientRegistration, JWK> jwkResolver;
+
+	private final Map<String, JwtEncoder> jwtEncoders = new ConcurrentHashMap<>();
+
+	private JwtCustomizer<T> jwtCustomizer = (request, headers, claims) -> {
+	};
+
+	/**
+	 * Constructs a {@code NimbusJwtClientAuthenticationCustomizer} using the provided
+	 * parameters.
+	 * @param jwkResolver the resolver that provides the {@code com.nimbusds.jose.jwk.JWK}
+	 * associated to the {@link ClientRegistration client}
+	 */
+	public NimbusJwtClientAuthenticationCustomizer(Function<ClientRegistration, JWK> jwkResolver) {
+		Assert.notNull(jwkResolver, "jwkResolver cannot be null");
+		this.jwkResolver = jwkResolver;
+	}
+
+	@Override
+	public void customize(T authorizationGrantRequest, HttpHeaders headers, MultiValueMap<String, String> parameters) {
+		Assert.notNull(authorizationGrantRequest, "authorizationGrantRequest cannot be null");
+		Assert.notNull(headers, "headers cannot be null");
+		Assert.notNull(parameters, "parameters cannot be null");
+
+		ClientRegistration clientRegistration = authorizationGrantRequest.getClientRegistration();
+		if (!ClientAuthenticationMethod.PRIVATE_KEY_JWT.equals(clientRegistration.getClientAuthenticationMethod())
+				&& !ClientAuthenticationMethod.CLIENT_SECRET_JWT
+						.equals(clientRegistration.getClientAuthenticationMethod())) {
+			return;
+		}
+
+		JWK jwk = this.jwkResolver.apply(clientRegistration);
+		if (jwk == null) {
+			OAuth2Error oauth2Error = new OAuth2Error(INVALID_KEY_ERROR_CODE,
+					"Failed to resolve JWK signing key for client registration '"
+							+ clientRegistration.getRegistrationId() + "'.",
+					null);
+			throw new OAuth2AuthorizationException(oauth2Error);
+		}
+
+		JwsAlgorithm jwsAlgorithm = resolveAlgorithm(jwk);
+		if (jwsAlgorithm == null) {
+			OAuth2Error oauth2Error = new OAuth2Error(INVALID_ALGORITHM_ERROR_CODE,
+					"Unable to resolve JWS (signing) algorithm from JWK associated to client registration '"
+							+ clientRegistration.getRegistrationId() + "'.",
+					null);
+			throw new OAuth2AuthorizationException(oauth2Error);
+		}
+
+		JoseHeader.Builder headersBuilder = JoseHeader.withAlgorithm(jwsAlgorithm);
+
+		Instant issuedAt = Instant.now();
+		Instant expiresAt = issuedAt.plus(Duration.ofSeconds(30));
```

Comment:
I increased the default to 60 secs. It could be further customized by the application via `JwtCustomizer` based on the requirements of the Authorization Server policy.